File: ngpy-accounts-manager-api/api/domain/services/budget_service.py
from datetime import date
from typing import List

from .transaction_service import TransactionService
from ..models import Budget, BudgetStatus, PeriodType
from ..search_request import FilterRequest, FilterOperator
from ...dbconnector.entities import BudgetDbo
from ...dbconnector.repositories import BudgetRepository
from ...mapping import Mapper
from ...modules.depynject import injectable


@injectable()
class BudgetService:
    """
    The budget service class that defines all business operations.
    """

    def __init__(self,
                 budget_repository: BudgetRepository,
                 object_mapper: Mapper,
                 transaction_service: TransactionService
                 ) -> None:
        """Constructor

        :param budget_repository: the budget repository
        :param object_mapper: the object mapper
        :param transaction_service: the transaction service
        """
        self.__repository = budget_repository
        self.__mapper = object_mapper
        self.__transaction_service = transaction_service

    def find_all(self, filter_request: FilterRequest) -> List[Budget]:
        """Gets a list of all budgets matching the provided filters.

        :param filter_request: the filter request
        :return: the list of all budgets
        """
        return self.__mapper.map_all(
            self.__repository.find_all(filter_request),
            Budget
        )

    def get_by_id(self, budget_id: int) -> Budget:
        """Gets a budget by its id.

        :param budget_id: the budget id
        :return: the budget
        """
        return self.__mapper.map(
            self.__repository.get_by_id(budget_id),
            Budget
        )

    def delete_budget(self, budget_id: int) -> None:
        """Deletes a budget by its id.

        :param budget_id: the budget id
        """
        self.__repository.delete_by_id(budget_id)

    def save_budget(self, budget: Budget) -> Budget:
        """Saves a given budget (create and update).

        :param budget: the budget to save
        :return: the saved budget
        """
        saved_budget = self.__repository.save_one(
            self.__mapper.map(budget, BudgetDbo)
        )
        return self.__mapper.map(saved_budget, Budget)

    def get_status_list(self,
                        status_date: date,
                        filter_request: FilterRequest
                        ) -> List[BudgetStatus]:
        """Gets a list of all budget status matching the provided filters.

        :param status_date: the end date
        :param filter_request: the filter request
        :return: the list of all budgets status
        """
        budgets = self.__repository.find_all(filter_request)
        status_list = []
        for budget in budgets:
            # Adding the label filters
            label_ids = list(
                map(lambda l: l.id, budget.labels)
            )
            transaction_filters = FilterRequest.all(
                FilterRequest.of('label_id', label_ids, FilterOperator.IN)
            )

            # Adding the account filters
            if budget.accounts is not None and len(budget.accounts) > 0:
                account_ids = list(
                    map(lambda a: a.id, budget.accounts)
                )
                transaction_filters = FilterRequest.all(
                    FilterRequest.of('account_id', account_ids, FilterOperator.IN),
                    transaction_filters
                )

            # Get the total spending on the period
            period_type = PeriodType.resolve(budget.period)
            if period_type == PeriodType.DAY:
                total = self.__get_daily_total(status_date, transaction_filters)
            elif period_type == PeriodType.MONTH:
                total = self.__get_monthly_total(status_date, transaction_filters)
            elif period_type == PeriodType.QUARTER:
                total = self.__get_quarterly_total(status_date, transaction_filters)
            else:
                total = self.__get_yearly_total(status_date, transaction_filters)

            # Create the status object and append to list
            status = BudgetStatus(budget, total)
            status_list.append(status)

        return status_list

    def __get_daily_total(self, status_date: date, transaction_filters: FilterRequest) -> float:
        """Gets the daily total for a given date matching the provided filters

        :param status_date: the date
        :param transaction_filters: the filters
        :return: the total
        """
        date_from = date(status_date.year, status_date.month, status_date.day)
        date_to = date(status_date.year, status_date.month, status_date.day + 1)
        filters = FilterRequest.all(
            FilterRequest.of('date_value', date_from, FilterOperator.GE),
            FilterRequest.of('date_value', date_to, FilterOperator.LT),
            transaction_filters
        )
        return self.__transaction_service.get_total(filters)

    def __get_monthly_total(self, status_date: date, transaction_filters: FilterRequest) -> float:
        """Gets the monthly total for a given date matching the provided filters

        :param status_date: the date
        :param transaction_filters: the filters
        :return: the total
        """
        date_from = date(status_date.year, status_date.month, 1)
        date_to = date(status_date.year, status_date.month + 1, 1)
        filters = FilterRequest.all(
            FilterRequest.of('date_value', date_from, FilterOperator.GE),
            FilterRequest.of('date_value', date_to, FilterOperator.LT),
            transaction_filters,
        )
        return self.__transaction_service.get_total(filters)

    def __get_quarterly_total(self, status_date: date, transaction_filters: FilterRequest) -> float:
        """Gets the quarterly total for a given date matching the provided filters

        :param status_date: the date
        :param transaction_filters: the filters
        :return: the total
        """
        quarter = int((status_date.month - 1) / 3) + 1
        start_month = (quarter - 1) * 3
        date_from = date(status_date.year, start_month, 1)
        if quarter == 4:
            date_to = date(status_date.year, quarter * 3, 1)
        else:
            date_to = date(status_date.year + 1, 1, 1)
        filters = FilterRequest.all(
            FilterRequest.of('date_value', date_from, FilterOperator.GE),
            FilterRequest.of('date_value', date_to, FilterOperator.LT),
            transaction_filters,
        )
        return self.__transaction_service.get_total(filters)

    def __get_yearly_total(self, status_date: date, transaction_filters: FilterRequest) -> float:
        """Gets the yearly total for a given date matching the provided filters

        :param status_date: the date
        :param transaction_filters: the filters
        :return: the total
        """
        date_from = date(status_date.year, 1, 1)
        date_to = date(status_date.year + 1, 1, 1)
        filters = FilterRequest.all(
            FilterRequest.of('date_value', date_from, FilterOperator.GE),
            FilterRequest.of('date_value', date_to, FilterOperator.LT),
            transaction_filters,
        )
        return self.__transaction_service.get_total(filters)

File: ngpy-accounts-manager-ui/src/app/core/services/domain/budget.service.ts
import { Injectable } from '@angular/core';
import { RqlService } from '../rql.service';
import { Observable } from 'rxjs';
import { Budget, BudgetStatus, KeyValue, Transaction } from '../../models/api.models';
import { DateService } from '../date.service';
import { BudgetRestService } from '../rest/budget-rest.service';
import { FilterOperator, FilterRequest } from '../../models/rql.models';
import { StatisticsRestService } from '../rest/statistics-rest.service';
import { TransactionsRestService } from '../rest/transactions-rest.service';

/**
 * The budget service
 */
@Injectable()
export class BudgetService {

  /**
   * Instantiates the service.
   *
   * @param budgetRestService the budget rest service
   * @param statisticsRestService the statistics rest service
   * @param transactionsRestService the transactions rest service
   * @param dateService the date service
   * @param rqlService the RQL service
   */
  constructor(private budgetRestService: BudgetRestService,
              private statisticsRestService: StatisticsRestService,
              private transactionsRestService: TransactionsRestService,
              private dateService: DateService,
              private rqlService: RqlService
  ) {
  }

  /**
   * Gets the details of a budget by its id.
   *
   * @param budgetId the budget id
   */
  getDetails(budgetId: number): Observable<Budget> {
    return this.budgetRestService.getOne(budgetId);
  }

  /**
   * Gets the status list for a given list of accounts, a year and a month.
   *
   * @param accounts the list of accounts
   * @param year the year
   * @param month the month
   */
  getStatusList(accounts: number[], year: number, month: number): Observable<BudgetStatus[]> {
    let filters: FilterRequest;
    // Adding the account filters
    if (accounts && accounts.length > 0) {
      const accountList = this.rqlService.formatList(accounts);
      filters = FilterRequest.of('accountId', accountList, FilterOperator.IN);
    }

    const statusDate = this.dateService.getPeriodStart(year, month);
    return this.budgetRestService.getStatusList(statusDate, filters);
  }

  /**
   * Gets the status history.
   *
   * @param dateFrom
   * @param dateTo
   * @param period
   * @param accounts
   * @param labels
   */
  getStatusHistory(
    dateFrom: Date,
    dateTo: Date,
    period: string,
    accounts: number[],
    labels: number[]
  ): Observable<KeyValue[]> {
    // Adding date and amount filters
    let filter = FilterRequest.all(
      FilterRequest.of('dateValue', this.dateService.format(dateFrom), FilterOperator.GE),
      FilterRequest.of('dateValue', this.dateService.format(dateTo), FilterOperator.LT)
    );

    // Adding the account filters
    if (accounts && accounts.length > 0) {
      const accountList = this.rqlService.formatList(accounts);
      filter = FilterRequest.all(
        FilterRequest.of('accountId', accountList, FilterOperator.IN),
        filter
      );
    }

    // Adding the label filters
    if (labels && labels.length > 0) {
      const accountList = this.rqlService.formatList(labels);
      filter = FilterRequest.all(
        FilterRequest.of('labelId', accountList, FilterOperator.IN),
        filter
      );
    }

    return this.statisticsRestService.getAggregation(period, filter);
  }

  /**
   * Gets the list of transactions.
   *
   * @param dateFrom
   * @param dateTo
   * @param accounts
   * @param labels
   */
  getTransactions(
    dateFrom: Date,
    dateTo: Date,
    accounts: number[],
    labels: number[]
  ): Observable<Transaction[]> {

    let filters = FilterRequest.all(
      FilterRequest.of('dateValue', this.dateService.format(dateFrom), FilterOperator.GE),
      FilterRequest.of('dateValue', this.dateService.format(dateTo), FilterOperator.LT)
    );
    if (accounts && accounts.length > 0) {
      filters = FilterRequest.all(
        FilterRequest.of('accountId', this.rqlService.formatList(accounts), FilterOperator.IN),
        filters
      );
    }
    if (labels && labels.length > 0) {
      filters = FilterRequest.all(
        FilterRequest.of('labelId', this.rqlService.formatList(labels), FilterOperator.IN),
        filters
      );
    }

    return this.transactionsRestService.getAll({ filter: filters });
  }

}
